src/sql/query.rs

```diff
@@ -1597,6 +1597,41 @@ fn plan_aggregate(ecx: &ExprContext, sql_func: &Function) -> Result<AggregateExp
             }
         }
     };
+    if let Some(filter) = &sql_func.filter {
+        // If a filter is present, as in
+        //
+        //     <agg>(<expr>) FILTER (WHERE <cond>)
+        //
+        // we plan it by essentially rewriting the expression to
+        //
+        //     <agg>(CASE WHEN <cond> THEN <expr> ELSE NULL)
+        //
+        // as aggregate functions ignore NULL. The only exception is `count(*)`,
+        // which includes NULLs in its count; we handle that specially by
+        // rewriting to:
+        //
+        //     count(CASE WHEN <cond> THEN TRUE ELSE NULL)
+        //
+        // (Note the `TRUE` in in place of `<expr>`.)
+        let cond = plan_expr(&ecx.with_name("FILTER"), filter, Some(ScalarType::Bool))?;
+        let cond_typ = ecx.column_type(&cond);
+        if cond_typ.scalar_type != ScalarType::Bool && cond_typ.scalar_type != ScalarType::Unknown {
+            bail!(
+                "WHERE expression in FILTER must have boolean type, not {:?}",
+                cond_typ
+            );
+        }
+        let expr_typ = ecx.scalar_type(&expr);
+        if func == AggregateFunc::CountAll {
+            func = AggregateFunc::Count;
+            expr = ScalarExpr::literal_true();
+        }
+        expr = ScalarExpr::If {
+            cond: Box::new(cond),
+            then: Box::new(expr),
+            els: Box::new(ScalarExpr::literal_null(expr_typ)),
```

Comment:
I don't know enough SQL to know, but is it correct that CountAll cannot have the `distinct` modifier set?

Reply:
It's supposed to be a syntax error, but it is apparently not in materialized! I'll file a bug.

Reply:
Is it also a syntax error to write `count(*) filter (where <expr>)`? That will result in potentially surprising behavior here. Also, unless I misunderstand, this seems to remove `CountAll` as a variant that we need to deal with downstream (e.g. in rendering); perhaps we could reflect that in the types.

Reply:
Sorry, reading more about postgres's take on the intended semantics, maybe it's all fine? I misunderstood the "can replace expr with ...". But, it does seem like `CountAll` probably doesn't ever need to exist? It doesn't seem like `count(*)` does anything other than `count(0)`, and perhaps should just get planned as that?

Reply:
Logically equivalent, I think, but `count(0)` will use an extra 9 bytes per row.

Reply:
In `reduce.rs`, do you mean? I think these all just get packed away in the diff field.

Reply:
I'm looking at https://github.com/MaterializeInc/materialize/blob/master/src/dataflow/render/reduce.rs#L183-L187 which will have to do extra row packing. Not a big deal, probably.

Reply:
Gotcha. That should just be transient, but 100% right that it will be created. We could certainly optimize that, but I'm guessing that constant expressions probably aren't common in other aggregations